[Re-land] Do not speculate InBounds if we see LoadFromHole exit

justinmichaud opened the pull request:
#### 14b4474c3e5c6155713e3e73b20913e55adadc5d
<pre>
[Re-land] Do not speculate InBounds if we see LoadFromHole exit
<a href="https://bugs.webkit.org/show_bug.cgi?id=303879">https://bugs.webkit.org/show_bug.cgi?id=303879</a>

Reviewed by Keith Miller.

This fixes an issue when landing <a href="https://commits.webkit.org/303355@main.">https://commits.webkit.org/303355@main.</a>

We must always refine the array mode if it is SelectUsingArguments,
even if we decide to add OutOfBounds.

Canonical link: <a href="https://commits.webkit.org/305377@main">https://commits.webkit.org/305377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99791ec27bd9ecafbacb20dd90484e20676b0f2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/91030 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105580 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/77041 "6 flakes 10 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/141001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8308 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123780 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86429 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7921 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5677 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6408 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/130019 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117319 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148838 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/136611 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10102 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/114006 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114318 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29082 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7851 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120059 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64826 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10148 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/38017 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/169327 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9879 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73716 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/44158 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10089 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
```diff
@@ -217,7 +217,10 @@ ArrayMode ArrayMode::refine(
-    
+
+    if (graph.hasExitSite(node->origin.semantic, LoadFromHole))
+        return ArrayMode(Array::Generic, action());
```
This seems like it would be bad, why would we not want to just fall back to OutOfBounds with the more accurate indexing type?

justinmichaud replied:
It is actually pretty rare from what I can tell, you need to get really unlucky with what the compiler sees to get an exit like this. Also, I wanted to avoid breaking things again.

I could factor out the code below this, and just tack on OutOfBounds if that looks better to you.

justinmichaud replied:
This turned out to be way simpler than I was imagining

kmiller68 reviewed (commented):
Question

kmiller68 reviewed (approved):
r=me if you update the commit message to the current change i.e. not say it's `Generic`

justinmichaud committed:
[Re-land] Do not speculate InBounds if we see LoadFromHole exit
https://bugs.webkit.org/show_bug.cgi?id=303879

Reviewed by Keith Miller.

This fixes an issue when landing https://commits.webkit.org/303355@main.

We must always refine the array mode if it is SelectUsingArguments,
even if we decide to add OutOfBounds.

Canonical link: https://commits.webkit.org/305377@main
